signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,314 @@
+    }
+
+    pub fn add_seed_addresses<T>(&mut self, seed_addresses: T) -> &mut Self
+    where
+        T: IntoIterator<Item = Multiaddr>,
+    {
+        self.config.seed_addresses.extend(seed_addresses);
+        tracing::debug!(addresses = ?self.config.seed_addresses, "added seed addresses");
+        self
+    }
+}
+
+impl NetworkBehaviour for Behavior {
+    type ConnectionHandler = libp2p::swarm::dummy::ConnectionHandler;
+    type ToSwarm = BootstrapEvent;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _peer: PeerId,
+        _local_addr: &Multiaddr,
+        _remote_addr: &Multiaddr,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        Ok(libp2p::swarm::dummy::ConnectionHandler)
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: libp2p::core::Endpoint,
+        _port_use: libp2p::core::transport::PortUse,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        Ok(libp2p::swarm::dummy::ConnectionHandler)
+    }
+
+    fn handle_pending_outbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _maybe_peer: Option<PeerId>,
+        addresses: &[Multiaddr],
+        effective_role: libp2p::core::Endpoint,
+    ) -> Result<Vec<Multiaddr>, libp2p::swarm::ConnectionDenied> {
+        tracing::debug!(?addresses, "handling pending outbound connection");
+        let addresses = addresses.to_vec();
+
+        // We're only interested in our outbound dialing activity. This _can_ be
+        // a listener if the swarm is attempting a hole-punch.
+        if !effective_role.is_dialer() {
+            return Ok(addresses);
+        }
+
+        for addr in &addresses {
+            if self.is_seed_address(addr) {
+                tracing::debug!(%addr, "attempting to dial seed address");
+            }
+        }
+
+        // Return the addresses as-is.
+        Ok(addresses)
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm) {
+        match event {
+            FromSwarm::ConnectionEstablished(e) => {
+                // We're only interested in our outbound activity.
+                if !e.endpoint.is_dialer() {
+                    return;
+                }
+
+                // If we've connected to ourselves, we don't need to do anything.
+                // This might happen if a node has its own address in the seed list.
+                if e.peer_id == self.local_peer_id() {
+                    return;
+                }
+
+                // Increment our total connection count. We're not interested
+                // in counting connections to ourselves.
+                self.connection_count = self.connection_count.saturating_add(1);
```

Comment:
Shouldn't this take into account only seed connections? Otherwise it could be possible that some subset of nodes are still connected one another but none is connected to any seed (creating a partition).

Reply:
No -- the idea with the seed nodes is that they're a well-known source of additional peers, but there should be no requirement that peers maintain active connections to the seed nodes.

I get what you're thinking, but that's largely due to the fact that we have exactly one seed node today -- we should change this so that we have at least 2 or 3 trusted signers who can commit to maintaining a static IP and be available for inbound connections, and have all signers add them to their seed list.

Also, implementing persistence of discovered peers to the DB would also alleviate any issues around seed nodes restarting, so they could attempt to re-connect to their most recently connected peers and re-discover nodes.